### PR TITLE
Propagate miette error info through FFI

### PR DESCRIPTION
**Pull request by cdisselkoen**
## Description of changes

FFI returns rich JSON structures for errors, rather than a simple string.

This diverges from the plan proposed in #757 in two ways: 
1. This PR actually specifies the miette JSON structure produced, rather than just using `serde_json::Value` in our interface types.  This makes it easier to integrate with the FFI (and write tests), and makes us responsible for updates if the `miette` JSON structure ever changes, but as an upside the code will complain louder and earlier if the `miette` JSON structure does change.
2. `ValidationWarning` and `ValidationError` have the same structure, so this PR proposes just having a single struct `ValidationError`.  Errors and warnings can still be distinguished from each other in two ways: (1) via `miette`'s `severity` field; and (2) because `ValidationAnswer::Success` keeps them in separate lists `validation_errors` and `validation_warnings`.

Somewhat orthogonally, this PR also includes a pass through all of the `cedar` repo removing occurrences to the outdated terminology "instantiation" in favor of "linking".  This can be separated into a different PR if that's preferable.

## Issue #, if available

Part of #757 

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).

I confirm that this PR (choose one, and delete the other options):

- [x] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)

### CODE REVIEW

**Comment by cdisselkoen** [comment]
for review:

1. the `REVIEW` comment in `cedar-wasm/src/validator.rs`
2. The code currently has two possible panics in `cedar-policy/src/ffi/utils.rs`.  Do we want to: (1) justify these with PANIC_SAFETY; (2) bite the bullet and make this a `TryFrom` instead of `From`, with some error type indicating it's a purely internal error; or (3) re-architect to not use `miette`'s JSON at all and just write our own serialization from `miette::Report` to a very similar JSON structure, that we can (a) control better and (b) know won't panic

**Comment by khieta** [comment]
When I last looked at the miette JSON format, it didn't seem like it was particularly well-documented (and thus maybe not widely used), so I'd lean towards a custom JSON format. We would presumably have a function that converts miette's format to ours, so that would be an indicator when something breaks.

**Comment by cdisselkoen** [comment]
Yes, it doesn't seem to be well-documented, and in fact I just looked at some outputs experimentally when crafting the structure in this PR.

File: cedar-wasm/src/validator.rs [hunk 1]
```diff
@@ -14,41 +14,61 @@
...
 use serde::{Deserialize, Serialize};
 use tsify::Tsify;
 use wasm_bindgen::prelude::*;
 
+/// `ffi::ValidationAnswer`, but with `serde(rename_all = "camelCase")`
+///
+/// REVIEW: do we want to just camelCase the fields in the normal
+/// `ValidationAnswer`, and adapt Java and others? It's confusing for
+/// Rust<->Wasm to use camelCased JSON field names but Rust<->Java to use
+/// snake_cased JSON field names
```
Comment (khieta) [inline cedar-wasm/src/validator.rs:27]: No strong feelings. Would it be weird if we set up serde `alias`es to support both snake and camel case (and strange combinations thereof)?

Comment (khieta) [reply cedar-wasm/src/validator.rs:27]: Various opinions on stack overflow: https://stackoverflow.com/questions/5543490/json-naming-convention-snake-case-camelcase-or-pascalcase

Comment (cdisselkoen) [reply cedar-wasm/src/validator.rs:27]: Serde `alias` is a good idea, but only helps for deserialization.  Serialization still has to pick one.

**Comment by cdisselkoen** [comment]
Pushed another rev of this PR, now uses our own custom error JSON format and doesn't touch miette's JSON handler at all.  Gives us maximum control over the JSON shape.  Allows us to have `message` include all the `source`s (which seems to be the pre-existing behavior).

File: cedar-wasm/src/validator.rs (hunk 1 above)
Comment (john-h-kastner-aws) [reply cedar-wasm/src/validator.rs:27]: aliasing does feel a little weird IMO. I prefer picking one convention for everything

Comment (cdisselkoen) [reply cedar-wasm/src/validator.rs:27]: TS folks seem to prefer camelCase in #805

Comment (Swolebrain) [reply cedar-wasm/src/validator.rs:27]: Yeah camelCase is definitely preferred in ts/js

Comment (cdisselkoen) [reply cedar-wasm/src/validator.rs:27]: #814 making the global change to camelCase in JSON throughout the `ffi` module

**Comment by cdisselkoen** [comment]
Updated this PR given #814.  Made the cedar-wasm changes a bunch simpler.

**Comment by cdisselkoen** [comment]
Updated this PR to account for the changes made in #824.  This PR does propose breaking changes to some of the choices made in #824:

* Errors can have multiple source locations, at least in the fullness of time.  (Cedar doesn't use this much currently, but `miette` supports it, and it seems like a good future-proofing measure to support it in the FFI now.)  This PR proposes to rename #824's `sourceLocation` to `sourceLocations` and make it a JSON array.
* Because there are many fields (up to 7) associated with each error now, this PR proposes to rename the main error message from `error` to `message`.
* This PR also introduces one more level of nesting for errors compared to #824 (which added `sourceLocation` as a peer of `error`): validation errors are now `{ policyId: ..., error: { message: ..., sourceLocations: [...], ...}`.  This could be flattened, but I thought it was valuable to have a consistent structure for errors which could be mirrored on the TS side, and the contents of `error` in this struct (`DetailedError` in the Rust in this PR) are used in many other places now, and there is not always a `policyId` associated with all of these errors (this struct is used for other kinds of errors as well, not just errors on policies).  I'm open to removing the nesting if yall think the flatter structure is more important than a consistent `DetailedError` type.
* `type: "error"` renamed to `type: "failure"`.  I think this is more clear, since the `"success"` case can still contain policy evaluation errors.  The top-level `Failure` is only for failures to parse or call the authorizer entirely.

**Comment by chrnorm** [comment]
Chiming in with some of the work I've been doing on our [GitHub Actions integration](https://github.com/common-fate/cedar-validate-action) - the error structure proposed here (`{ policyId: ..., error: { message: ..., sourceLocations: [...], ...}`) would be very easy to work with and I am in favor of it.

File: cedar-policy/src/ffi/mod.rs [hunk 2]
```diff
@@ -14,9 +14,11 @@
  * limitations under the License.
  */
 
+#![cfg_attr(feature = "wasm", allow(non_snake_case))]
```
Comment (khieta) [inline cedar-policy/src/ffi/mod.rs:17]: Out of curiosity: what is this for? I thought wasm preferred camelCase to snake_case?

File: cedar-policy/src/ffi/utils.rs [hunk 3]
```diff
@@ -17,12 +17,134 @@
...
+        }
+    }
+}
+
+impl<'a, E: miette::Diagnostic + ?Sized> From<&'a E> for DetailedError {
+    fn from(diag: &'a E) -> Self {
+        Self {
+            message: {
+                let mut s = diag.to_string();
+                let mut source = diag.source();
```
Comment (khieta) [inline cedar-policy/src/ffi/utils.rs:117]: What is `.source()`? Not related to source locations?

**Review by khieta — APPROVED** [review]
Looking good! The only comments I left were clarifying questions for my own understanding.

Also, it's a little late in the process for me to mention this, but I'd prefer if the "instantiate" -> "link" changes were in a separate PR 🙂

**Review by john-h-kastner-aws — APPROVED** [review]
Looks good.

File: cedar-policy/src/ffi/mod.rs (hunk 2 above)
Comment (cdisselkoen) [reply cedar-policy/src/ffi/mod.rs:17]: Wasm does prefer camelCase.  Including, some of tsify's macros expand to functions with camelCase names, which clippy doesn't like.  This `allow` avoids warnings for these autogenerated camelCased functions.

File: cedar-policy/src/ffi/utils.rs (hunk 3 above)
Comment (cdisselkoen) [reply cedar-policy/src/ffi/utils.rs:117]: It's a method on the [`Error` trait in std](https://doc.rust-lang.org/std/error/trait.Error.html).  I'm not sure why, but I needed to add these `:` and `source()` in order to preserve previous behavior (where this is the string message that was previously passed in the FFI).

**Comment by cdisselkoen** [comment]
Split the "instantiate" -> "link" changes into #826

**Comment by cdisselkoen** [comment]
Now that #826 has been merged, I've rebased this PR on `main` and it is ready for re-review

**Review by khieta — APPROVED** [review]

**Pull request merged**